src/main/java/org/opensearch/neuralsearch/executors/HybridQueryExecutorCollector.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.executors;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import java.util.function.Function;
+
+/**
+ * {@link HybridQueryExecutorCollector} is a generic Collector used by Hybrid Search Query during
+ * Query phase to parallelize sub query's action to improve latency
+ */
+@RequiredArgsConstructor(staticName = "newCollector")
+public final class HybridQueryExecutorCollector<I, R> {
```

Comment:
Is `HybridQueryExecutorCollector` thread safe?

Reply:
There are no setters and only way to update result is by calling collect method. Do you have any particular concerns?

Reply:
Since `collect()` could set `result`, if `collect()` and `getResults()` are called asynchronously, technically, there could be race condition, right?

Reply:
You are right. Currently getResult() is called from collector manager, where the contract is that it can be called only after all collectors are finished collection. This was taken care in the implementation. I can add this note in collector  as well. i am reluctant to synchronize to result, since it will add additional latency for scenario that is not possible at this moment. Do you have any other suggestions?

Reply:
Adding a comment is a good intention. you won't be able to enforce it. You are relying on the hope that the code doesn't change in future to hit that race condition so there will always be a risk.

If you really want to enforce; your options here would be to synchronize or move away from `result collection` to `returning results` as pointed out in [this thread](https://github.com/opensearch-project/neural-search/pull/749/files#r1608714432)

Reply:
> i am reluctant to add synchronize to result variable, since it will add additional latency for scenario that is not possible at this moment

Actually, the overhead of synchronization is way less than we thought, here are some synchronization benchmark articles I found, 
- https://baptiste-wicht.com/posts/2010/09/java-synchronization-mutual-exclusion-benchmark.html
- https://isuru-perera.blogspot.com/2016/05/benchmarking-java-locks-with-counters.html
One operation of "synchronized" is < 1 micro second, considering the time spending on actual query logic with intensive I/O, this is negligible and you have simple implementation with concurrent situation covered.

Reply:
@shatejas @chishui Fair enough. I am changing to synchronized, however while benchmarking i am planning to compare with and without synchronization to see that it is not adding any latency to this block.